specs/_features/eip7732/beacon-chain.md

```diff
@@ -0,0 +1,796 @@
+#### `PayloadAttestation`
+
+```python
+class PayloadAttestation(Container):
+    aggregation_bits: Bitvector[PTC_SIZE]
+    data: PayloadAttestationData
+    signature: BLSSignature
+```
+
+#### `PayloadAttestationMessage`
+
+```python
+class PayloadAttestationMessage(Container):
+    validator_index: ValidatorIndex
+    data: PayloadAttestationData
+    signature: BLSSignature
+```
+
+#### `IndexedPayloadAttestation`
+
+```python
+class IndexedPayloadAttestation(Container):
+    attesting_indices: List[ValidatorIndex, PTC_SIZE]
+    data: PayloadAttestationData
+    signature: BLSSignature
+```
+
+#### `SignedExecutionPayloadHeader`
+
+```python
+class SignedExecutionPayloadHeader(Container):
+    message: ExecutionPayloadHeader
+    signature: BLSSignature
+```
+    
+#### `ExecutionPayloadEnvelope`
+
+```python
+class ExecutionPayloadEnvelope(Container):
+    payload: ExecutionPayload
+    builder_index: ValidatorIndex
+    beacon_block_root: Root
+    blob_kzg_commitments: List[KZGCommitment, MAX_BLOB_COMMITMENTS_PER_BLOCK]
+    payload_withheld: boolean
+    state_root: Root
+```
+
+#### `SignedExecutionPayloadEnvelope`
+
+```python
+class SignedExecutionPayloadEnvelope(Container):
+    message: ExecutionPayloadEnvelope
+    signature: BLSSignature
+```
+
+### Modified containers
+
+#### `BeaconBlockBody`
+
+**Note:** The Beacon Block body is modified to contain a `Signed ExecutionPayloadHeader`. The containers `BeaconBlock` and `SignedBeaconBlock` are modified indirectly.
+
+```python
+class BeaconBlockBody(Container):
+    randao_reveal: BLSSignature
+    eth1_data: Eth1Data  # Eth1 data vote
+    graffiti: Bytes32  # Arbitrary data
+    # Operations
+    proposer_slashings: List[ProposerSlashing, MAX_PROPOSER_SLASHINGS]
+    attester_slashings: List[AttesterSlashing, MAX_ATTESTER_SLASHINGS]
+    attestations: List[Attestation, MAX_ATTESTATIONS]
+    deposits: List[Deposit, MAX_DEPOSITS]
+    voluntary_exits: List[SignedVoluntaryExit, MAX_VOLUNTARY_EXITS]
+    sync_aggregate: SyncAggregate
+    # Execution
+    # Removed execution_payload [Removed in EIP-7732]
+    # Removed blob_kzg_commitments [Removed in EIP-7732]
+    bls_to_execution_changes: List[SignedBLSToExecutionChange, MAX_BLS_TO_EXECUTION_CHANGES]
+    # PBS
+    signed_execution_payload_header: SignedExecutionPayloadHeader   # [New in EIP-7732]
+    payload_attestations: List[PayloadAttestation, MAX_PAYLOAD_ATTESTATIONS]  # [New in EIP-7732]
+```
+
+#### `ExecutionPayloadHeader`
+
+**Note:** The `ExecutionPayloadHeader` is modified to only contain the block hash of the committed `ExecutionPayload` in addition to the builder's payment information, gas limit and KZG commitments root to verify the inclusion proofs. 
+
+```python
+class ExecutionPayloadHeader(Container):
+    parent_block_hash: Hash32
+    parent_block_root: Root
+    block_hash: Hash32
+    gas_limit: uint64
+    builder_index: ValidatorIndex
+    slot: Slot
+    value: Gwei
+    blob_kzg_commitments_root: Root
```

Comment:
Couldn't `blob_kzg_commitments_root` also be delayed until after `process_execution_payload`?

Reply:
I think in principle it could, the bid may not even commit a payload has even in principle. But I went with the minimal modification to the current consensus algo. Can switch to that if required. I don't see why this causes yunk though, it's just in the header that it's last committed. It doesn't appear anywhere else.